### PR TITLE
[翻译完毕] data layout

### PR DESCRIPTION
**Pull request by isdanni**
### Related issue
 #30 

### Not sure about following translations
- most `naive` data layout => 朴素
- low access locality => 低效的访问位置
- Struct-fors on advanced (dense) data layouts => 对高级（密集）数据布局进行for构造

Also, I added a ref link for TLB and translated to `转址旁路缓存`, other terms people also use include 页表缓存,  转译后备缓冲区; I checked on Zhihu, seems like there are larger group using `转址旁路缓存` LOL;

### CODE REVIEW

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 1]
```diff
@@ -3,69 +3,82 @@
...
-msgstr ""
+msgstr "高级数据布局"
 
 #: ../../data_layout.rst:4
 msgid ""
 "Memory layout is key to performance, especially for memory-bound "
 "applications. A carefully designed data layout can significantly improve "
 "cache/TLB-hit rates and cacheline utilization."
 msgstr ""
+"内存布局是性能的关键，特别是对于绑定内存的应用程序而言。 精心设计的数据布"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"内存布局是性能的关键，特别是对于内存密集型的应用程序而言。 精心设计的数据布"
```
See [CPU-bound vs I/O-bound](https://blog.csdn.net/u010229420/article/details/51538039).

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 2]
```diff
@@ -3,69 +3,82 @@
...
-"specifying ``shape`` when creating tensors using "
-"``ti.var/Vector/Matrix``), and then migrate to more advanced layouts "
-"using the ``ti.root.X`` syntax."
+"specifying ``shape`` when creating tensors using ``ti.var/Vector/"
+"Matrix``), and then migrate to more advanced layouts using the ``ti.root."
+"X`` syntax."
 msgstr ""
+"我们建议从默认布局规范开始（使用 ``ti.var/Vector/Matrix`` 创建张量时，只"
+"需指定 ``shape`` 即可），然后再使用 ``ti.root.X`` 句法迁移到更高级的布"
+"局。"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"我们建议从默认布局规范开始（在 ``ti.var/Vector/Matrix`` 中指定 ``shape`` 来创建张量）"
"，然后再迁移到用 ``ti.root.X`` 句法的更高级布局。"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 3]
```diff
@@ -3,69 +3,82 @@
...
+"and computer architecture."
 msgstr ""
+"Taichi将算法与数据布局分离，并且Taichi编译器会自动优化特定数据布局上的数"
+"据访问。 这些Taichi功能使程序员可以快速尝试不同的数据布局，并找出针对特定"
+"任务和计算机体系结构最有效的布局。"
 
 #: ../../data_layout.rst:16
 msgid "The default data layout using ``shape``"
-msgstr ""
+msgstr "使用 ``shape`` 的默认数据布局"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "使用 ``shape`` 时默认的数据布局"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 4]
```diff
@@ -3,69 +3,82 @@
...
+"data layout"
+msgstr "默认情况下，分配``ti.var``时，它遵循最朴素的数据布局"
 
 #: ../../data_layout.rst:20
 msgid ""
 "val = ti.var(ti.f32, shape=(32, 64, 128))\n"
 "# C++ equivalent: float val[32][64][128]"
 msgstr ""
+"val = ti.var(ti.f32, shape=(32, 64, 128))\n"
+"# C++ 当量: float val[32][64][128]"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"# 在 C++ 中相当于: float val[32][64][128]"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 5]
```diff
@@ -3,69 +3,82 @@
...
 
 #: ../../data_layout.rst:25
 msgid ""
 "Or equivalently, the same data layout can be specified using advanced "
 "`data layout description`:"
-msgstr ""
+msgstr "或者等效地，可以使用高级``数据布局描述``来指定相同的数据布局："
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "同样的数据布局也可以用高级数据布局来描述："
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 6]
```diff
@@ -74,6 +87,10 @@ msgid ""
 "# Specify the shape and layout\n"
 "ti.root.dense(ti.ijk, (32, 64, 128)).place(val)"
 msgstr ""
+"#创建全局张量\n"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"# 创建全局张量\n"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 7]
```diff
@@ -84,65 +101,80 @@ msgid ""
...
+"但是，对于计算机图形任务而言，这种数据布局通常不是最理想的。 例如，``val "
+"[i, j, k]`` 和 ``val [i + 1, j, k]`` 彼此之间的距离非常远 (``32 KB``)，并"
+"导致在某些计算任务下低效的访问位置。 具体来说，在诸如纹理三线性插值之类的"
+"任务中，这两个元素甚至不在同一 ``4KB`` 页面内，从而产生了巨大的缓存/转址"
+"旁路缓存压力。"
 
 #: ../../data_layout.rst:41
 msgid "Advanced data layout specification"
-msgstr ""
+msgstr "先进的数据布局规范"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "高级数据布局的指定"
```
Let's keep `advanced` -> `高级` throughout the doc.

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 8]
```diff
@@ -84,65 +101,80 @@ msgid ""
...
+"probability ``val[i, j, k]`` and its neighbours are close to each other "
+"(i.e., in the same cacheline or memory page)."
 msgstr ""
+"这会在 ``4x4x4`` 模块中组织 `` val``，因此很有可能 ``val [i, j, k]`` 及其"
+"邻居彼此靠近（即，在同一高速缓存行或内存页中） ）。"
 
 #: ../../data_layout.rst:53
 msgid "Examples"
-msgstr ""
+msgstr "例子"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "示例"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 9]
```diff
@@ -84,65 +101,80 @@ msgid ""
...
 "A = ti.var(ti.f32)\n"
 "ti.root.dense(ti.ji, (256, 256)).place(A) # Note ti.ji instead of ti.ij"
 msgstr ""
+"A = ti.var(ti.f32)\n"
+"ti.root.dense(ti.ji, (256, 256)).place(A) # 注意 ti.ji 而不是ti.ij"
 
 #: ../../data_layout.rst:69
 msgid "`8x8` blocked 2D array of size `1024x1024`"
-msgstr ""
+msgstr "`8x8` 阻止了大小为 `1024x1024` 的二维数组"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "按 `8x8` 分块，大小为 `1024x1024` 的二维数组"
```
Here `block` means `块` and `blocked` means `按……分块`. A similar example is:
`a tensor shaped (2, 3)` -> `以 (2, 3) 为形状的张量`.

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 10]
```diff
@@ -167,39 +205,52 @@ msgid ""
...
+"vel = ti.Vector(3, dt=ti.f32)\n"
+"for i in range(3):\n"
+"  ti.root.dense(ti.i, 1024).place(pos(i))\n"
+"for i in range(3):\n"
+"  ti.root.dense(ti.i, 1024).place(vel(i))"
 
 #: ../../data_layout.rst:100
 msgid "Struct-fors on advanced (dense) data layouts"
-msgstr ""
+msgstr "对高级（密集）数据布局进行for构造"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "对高级（密集）数据布局进行结构 for 循环"
```
`struct-for` -> `结构 for 循环`, please check out https://github.com/taichi-dev/taichi-docs-zh-cn/wiki/常用专有词汇对照表.

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 11]
```diff
@@ -167,39 +205,52 @@ msgid ""
...
 
 #: ../../data_layout.rst:102
 msgid ""
 "Struct-fors on nested dense data structures will automatically follow "
 "their data order in memory. For example, if 2D scalar tensor ``A`` is "
 "stored in row-major order,"
 msgstr ""
+"嵌套密集数据结构上的结构fors将自动遵循其在内存中的数据顺序。 例如，如果二"
+"维标量张量 ``A`` 以行优先顺序存储，"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"嵌套密集数据结构上的结构 for 循环将自动遵循其在内存中的数据顺序。 例如，如果二"
"维标量张量 ``A`` 以行优先顺序存储，"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 12]
```diff
@@ -167,39 +205,52 @@ msgid ""
...
+"for i, j in A:\n"
+"  A[i, j] += 1"
 
 #: ../../data_layout.rst:109
 msgid ""
 "will iterate over elements of ``A`` following row-major order. If ``A`` "
 "is column-major, then the iteration follows the column-major order."
 msgstr ""
+"将按照行优先顺序对A的元素进行迭代。 如果 ``A`` 是列主序，则迭代遵循列主"
+"序。"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]: `-major` -> `主序` or `优先`? Please choose one and keep it constant.
```suggestion
"将按照行优先的顺序对A的元素进行遍历。 如果 ``A`` 是列优先，则按照列优先进行遍历。"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 13]
```diff
@@ -167,39 +205,52 @@ msgid ""
...
 
 #: ../../data_layout.rst:111
 msgid ""
 "If ``A`` is blocked, the iteration will happen within each block first. "
 "This maximizes memory bandwidth utilization in most cases."
 msgstr ""
+"如果 ``A`` 被阻止，则迭代将首先在每个块内发生。 在大多数情况下，这可以最"
+"大化内存带宽利用率。"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"如果 ``A`` 是分块的，则迭代将首先在每个块内发生。 在大多数情况下，这可以最"
"大化内存带宽利用率。"
```
Again, `blocked` is `noun + ed`, not `verb + ed`. `block` -> `块`.

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 14]
```diff
@@ -167,39 +205,52 @@ msgid ""
...
 
 #: ../../data_layout.rst:113
 msgid ""
 "Struct-fors on sparse tensors follows the same philosophy, and will be "
 "discussed further in :ref:`sparse`."
-msgstr ""
-
+msgstr "稀疏张量的for构造遵循相同的原理，将在 :ref:`sparse` 中进一步讨论。"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
msgstr "稀疏张量的结构 for 循环遵循相同的原理，将在 :ref:`sparse` 中进一步讨论。"
```

File: locales/zh_CN/LC_MESSAGES/data_layout.po [hunk 15]
```diff
@@ -84,65 +101,80 @@ msgid ""
 "trilinear interpolation, the two elements are not even within the same "
 "``4KB`` pages, creating a huge cache/TLB pressure."
 msgstr ""
+"但是，对于计算机图形任务而言，这种数据布局通常不是最理想的。 例如，``val "
+"[i, j, k]`` 和 ``val [i + 1, j, k]`` 彼此之间的距离非常远 (``32 KB``)，并"
+"导致在某些计算任务下低效的访问位置。 具体来说，在诸如纹理三线性插值之类的"
+"任务中，这两个元素甚至不在同一 ``4KB`` 页面内，从而产生了巨大的缓存/转址"
```
Comment (archibate) [inline locales/zh_CN/LC_MESSAGES/data_layout.po]:
```suggestion
"任务中，这两个元素甚至不在同一 ``4KB`` 的页面内，从而产生巨大的缓存/转址"
```

**Review by archibate — COMMENTED** [review]
As it always, thank for your hard work!

File: locales/zh_CN/LC_MESSAGES/data_layout.po (hunk 12 above)
Comment (isdanni) [reply locales/zh_CN/LC_MESSAGES/data_layout.po]: Will stick to 优先, thanks!

**Comment by isdanni** [comment]
Updated! Thanks for the review, def learned a bunch!

**Review by archibate — APPROVED** [review]
Great! (Btw, you can use the button `Add suggestion to batch` in `Files changed` tab to merge many suggestions in one commit)

**Pull request merged**